### PR TITLE
add @torch.jit.script, @torch.jit.compile, torch.jit.CompilationUnit(str)

### PR DESCRIPTION
**Pull request by zdevito**
This PR just exposes our already existing functionality using annotations and a compilation unit object.
@apaszke  @jamesr66a

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -1348,7 +1347,7 @@ def func(x) -> (head):
         '''
         x = Variable(torch.rand(10).float(), requires_grad=True)
         outputs = x[:5]
-        self.checkScript(script, [x], [outputs], False)
+        self.checkScript(script, [x], outputs, False)
```
Comment (jamesr66a) [inline test/test_jit.py]: Why was this changed?

File: torch/csrc/jit/init.cpp [hunk 2]
```diff
@@ -119,17 +113,25 @@ void initJITBindings(PyObject *module) {
...
           }),
           py::arg("graph"),
           py::arg("optimize") = true)
-      .def("__call__", [](GraphExecutor& ge, py::args args) {
+      .def("__call__", [](GraphExecutor& ge, py::args args) -> py::object {
         auto inputs = createVariableTensorList(args);
         auto outputs = ge.run(std::move(inputs));
         // if we don't tell pybind these are variables it chokes on the
         // conversion.
         // TODO: fix conversions to be sane and make sure this works.
```
Comment (jamesr66a) [inline torch/csrc/jit/init.cpp]: Is this comment still relevant?

**Review by jamesr66a — APPROVED** [review]
coolcoolcool

File: test/test_jit.py [hunk 3]
```diff
@@ -1534,23 +1533,45 @@ def test_if_while(a, b) -> (c):
...
+        cu2 = torch.jit.CompilationUnit('''
         def test_ternary(a, b) -> (c):
             c = 3
             c = a + b if a > 3 else b
         ''')
         self.assertEqual(
-            str(cu.get_graph('test_ternary_control')),
-            str(cu2.get_graph('test_ternary')),
+            str(cu.cu.get_graph('test_ternary_control')),
+            str(cu2.cu.get_graph('test_ternary')),
```
Comment (apaszke) [inline test/test_jit.py]: Why do we have to index it twice now?

File: test/test_jit.py (hunk 1 above)
Comment (apaszke) [reply test/test_jit.py]: I think it's because GraphExecutors used to return tuples, but now return a single Variable, or a tuple of Variables that has at lest 2 elements.

**Review by apaszke — APPROVED** [review]
LGTM. One minor question

File: torch/csrc/jit/init.cpp (hunk 2 above)
Comment (zdevito) [reply torch/csrc/jit/init.cpp]: kind of -- we have to do the static_cast to Variable otherwise pybind tries to convert the tensors, and at::Tensor conversion doesn't work if the tensor is a Variable. I'd like to be able to drop the static_cast's but we are waiting for tensor/variable to merge to make it happen.

File: test/test_jit.py (hunk 3 above)
Comment (zdevito) [reply test/test_jit.py]: The frontend-facing CompilationUnit doesn't expose graphs, just the GraphExecutor objects that run the graphs. I just reached into the members of CompilationUnit here to get to the underlying C struct so we can check graph equality. If in the future we see the need for users to call get_graph, we can create a proxy method.

Comment (apaszke) [reply test/test_jit.py]: Oh I somehow entirely missed the fact that you added a Python class for it...

File: test/test_jit.py [hunk 4]
```diff
@@ -287,7 +287,7 @@ def f(x, y):
             return z1 * z2
         x = Variable(torch.randn(4, 4).float().cuda())
         y = Variable(torch.randn(4, 4).float().cuda())
-        trace, _ = torch.jit.trace(f, (x, y), nderivs=0)
+        trace, _ = torch.jit.create_trace(f, (x, y), nderivs=0)
```
Comment (apaszke) [inline test/test_jit.py]: Maybe we should call it `get_trace_graph` or sth? Just to make it clear it will give you a graph, and uses tracing, but is not connected to `trace` in any way

**Review by apaszke — APPROVED** [review]

File: test/test_jit.py (hunk 4 above)
Comment (zdevito) [reply test/test_jit.py]: sounds good.

**Comment by zdevito** [comment]
test failures appeared spurious, merging so the next set of PRs based on this can go in.
